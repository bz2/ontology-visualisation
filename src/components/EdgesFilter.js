/* eslint react/no-array-index-key:0 */
import {
  useState,
} from 'react'
import { connect } from 'redux-zero/react'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'
import { InputText } from 'primereact/inputtext'
import { Dropdown } from 'primereact/dropdown'
import { Button } from 'primereact/button'
import actions from '../store/actions'
import { SIDEBAR_VIEW_EDGES_FILTER } from '../constants/views'
import setFilteredEdges from '../utils/edgesFilter/setFilteredEdges'

const EdgesFilter = ({
  setStoreState,
  annotationProperties,
  addToObject
}) => {
  const { t } = useTranslation()

  const defaultEdgeFilter = {
    property: '',
    value: '',
  }

  const checkEmptyRow = (filters) => filters.filter((filter) => filter.property === '' && filter.value === '').length > 0

  const [edgesFilters, setEdgesFilters] = useState([defaultEdgeFilter])

  return (
    <>
      <div className="sidebar-main-title">
        {t(SIDEBAR_VIEW_EDGES_FILTER)}
      </div>

      <div className="edges-filter">
        <div>{t('filterEdgesByArbitraryPropsCombination')}</div>

        <div
          className="edges-filter-selection"
        >
          {
            edgesFilters.map(
              (edgefilter, index) => {
                const selectId = `edges-filter-property-${index}`
                const inputTextId = `edges-filter-value-${index}`

                return (
                  <div
                    className="edges-filter-selection-row"
                    key={`edges-filter-${index}`}
                  >
                    {
                      edgesFilters.length > 1 && (
                      <div className="p-field remove-button p-col-12 p-md-3">
                        <Button
                          icon="pi pi-times"
                          className="p-button-rounded p-button-danger"
                          tooltip={t('removeFilter')}
                          onClick={() => {
                            const newEdgeFilter = edgesFilters.slice()

                            newEdgeFilter.splice(index, 1)

                            if (!checkEmptyRow(newEdgeFilter)) {
                              newEdgeFilter.push(JSON.parse(JSON.stringify(defaultEdgeFilter)))
                            }

                            setEdgesFilters(newEdgeFilter)
                          }}
                        />
                      </div>
                      )
                    }

                    <div className="p-field p-col-12 p-md-3">
                      <label htmlFor={selectId}>{t('selectProperty')}</label>
                      <Dropdown
                        id={selectId}
                        value={edgesFilters[index].property}
                        options={annotationProperties}
                        filter
                        onChange={(e) => {
                          const newFilter = {
                            ...edgesFilters[index],
                            property: e.target.value.id
                          }

                          let newEdgesFilters = [
                            ...edgesFilters.slice(0, index),
                            newFilter,
                            ...edgesFilters.slice(index + 1),
                          ]

                          if (!checkEmptyRow(newEdgesFilters)) {
                            newEdgesFilters = [
                              ...newEdgesFilters,
                              JSON.parse(JSON.stringify(defaultEdgeFilter))
                            ]
                          }

                          setEdgesFilters(newEdgesFilters)
                        }}
                        className="m-t-10"
                        placeholder={t('selectProperty')}
                      />
                    </div>

                    <div className="p-field p-col-12 p-md-3 m-t-20">
                      <label htmlFor={inputTextId}>{t('searchString')}</label>
                      <InputText
                        id={inputTextId}
                        value={edgesFilters[index].value}
                        placeholder={t('searchString')}
                        onChange={(e) => {
                          const newFilter = {
                            ...edgesFilters[index],
                            value: e.target.value
                          }

                          setEdgesFilters([
                            ...edgesFilters.slice(0, index),
                            newFilter,
                            ...edgesFilters.slice(index + 1)
                          ])
                        }}
                        className="m-t-10"
                      />
                    </div>
                  </div>
                )
              }
            )
          }
        </div>

        <Button
          tooltip={t('showFilteredEdges')}
          className="edges-filter-button m-t-30"
          disabled={edgesFilters.length < 2}
          icon="pi pi-chevron-right"
          iconPos="right"
          label={t('show')}
          onClick={() => setFilteredEdges({
            setStoreState,
            edgesFilters,
            addToObject
          })}
        />
      </div>
    </>
  )
}

EdgesFilter.propTypes = {
  setStoreState: PropTypes.func.isRequired,
  annotationProperties: PropTypes.arrayOf(PropTypes.shape).isRequired,
  addToObject: PropTypes.func.isRequired,
}

const mapToProps = ({
  annotationProperties
}) => ({
  annotationProperties
})

export default connect(
  mapToProps,
  actions
)(EdgesFilter)
