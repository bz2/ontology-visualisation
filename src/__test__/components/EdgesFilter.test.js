import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import EdgesFilter from '../../components/EdgesFilter'

const setup = () => {
  const props = {
    edgesProperties: [{
      property: 'rdfsLabel',
      value: 'is composed of'
    }],
    updateGraphData: jest.fn(),
    setStoreState: jest.fn(),
  }

  const component = shallow(<EdgesFilter {...props} />)

  return {
    component,
    props
  }
}

describe('EdgesFilter', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot ', () => {
    const {
      component
    } = setup()

    expect(toJson(component)).toMatchSnapshot()
  })
})
