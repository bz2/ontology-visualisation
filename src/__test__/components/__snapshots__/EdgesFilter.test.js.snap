// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgesFilter should match snapshot  1`] = `
<Connect
  actions={
    Object {
      "addToArray": [Function],
      "removeFromArray": [Function],
      "removeFromObject": [Function],
      "resetSelectedEdges": [Function],
      "resetSelectedNodes": [Function],
      "setStoreState": [Function],
      "updateGraphData": [Function],
    }
  }
  edgesProperties={
    Array [
      Object {
        "property": "rdfsLabel",
        "value": "is composed of",
      },
    ]
  }
  mapToProps={[Function]}
  setStoreState={[MockFunction]}
  updateGraphData={[MockFunction]}
>
  <Component />
</Connect>
`;
